src/resolvelib/resolvers.py

```diff
@@ -302,6 +302,9 @@ def _backjump(self, causes):
 
             # Ensure to backtrack to a state that caused the incompatibility
             incompatible_state = False
+            incompatible_deps = {
+                self._p.identify(c.parent) for c in causes
```

Comment:
I am a still confused why it is `c.parent` not `c.requirement`. Let me show in an example, in the example, capitals are *candidates* while lowercases are *requirements*

```
State 1: A1 -> b1
State 2: C1 -> d1
State 3: E1 -> b2 and b1 conflicts with b2
```
Where `causes = [(b1, A1), (b2, E1)]`
We want to backjump to State 1, given:
```
get_dependencies(A1) = [b1]
```
this can be done by comparing the causes with the dependencies, shouldn't we use the .requirement attribute?

Reply:
The actual example I've been working with has 3 requirements:

1. boto3
2. s3fs
3. seaborn

So the states are the following:
```
State 1: NONE -> boto3
State 2: NONE -> s3fs
State 3: NONE -> seaborn
State 4: BOTO3 -> botocore
...
State 7: S3FS -> aiobotocore
...
State 11: SEABORN -> matplotlib
```

And `causes = [(botocore, BOTO3), (botocore, aiobotocore)]`

So, the incompatibility is caused by a dependency of one requirement and a transitive dependency of another requirement.

The function `get-dependencies` returns only the direct dependencies, so basing this on requirements only would cause an infinite loop.
Does it make sense?

Reply:
IIUC, `causes = [(botocore, BOTO3), (botocore, aiobotocore)]`, so we want to backjump to STATE 4 and mark BOTO3 as incompatible. To achieve this, we start from the **requirement name**(1) of causes, which is `botocore`, and traverse back the state stack, for each state, we call `get_dependencies()` on the new pin and see if the dependencies overlap with (1). So `get_dependencies(BOTO3) = [botocore]`, then we stop at STATE 4 and pop it from the stack.

If using `c.parent`, then (1) becomes [BOTO3, aiobotocore], how can we find the correct state?

Reply:
Oh, you want to backjump to State 7, then it makes sense

Reply:
Please also fix the lint errors to see if the CI is fine. And I will leave it to the other reviewers to decide whether this fix is valid. Thanks.

Reply:
The state `AIOBOTOCORE -> botocore` does not exist, because a state without a pin is not added to the states (see [here](https://github.com/sarugaku/resolvelib/blob/7b66e2da3c1d0189d0ccd85b68bb761d843754b1/src/resolvelib/resolvers.py#L428))

I tested with `c.requirement` and the code crashes.

One solution would be to append the state, even if unpinned, and remove [this line](https://github.com/sarugaku/resolvelib/blob/7b66e2da3c1d0189d0ccd85b68bb761d843754b1/src/resolvelib/resolvers.py#L297), any thoughts?